### PR TITLE
[SPARK-25855][CORE] Don't use erasure coding for event logs by default

### PR DESCRIPTION
**Pull request by squito**
## What changes were proposed in this pull request?

This turns off hdfs erasure coding by default for event logs, regardless of filesystem defaults.  Because this requires apis only available in hadoop 3, this uses reflection.  EC isn't a very good choice for event logs, as hflush() is a no-op, and so updates to the file are not visible for a long time.  This can still be configured by setting "spark.eventLog.allowErasureCoding=true", which will use filesystem defaults.

## How was this patch tested?

deployed a cluster with the changes with HDFS EC on.  By default, event logs didn't use EC, but configuration still would allow EC.  Also tried writing to the local fs (which doesn't support EC at all) and things worked fine.

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -761,6 +761,17 @@ Apart from these, the following properties are also available, and may be useful
...
   </td>
 </tr>
+<tr>
+  <td><code>spark.eventLog.allowErasureCoding</code></td>
+  <td>false</td>
+  <td>
+    Whether to allow event logs to use erasure coding, or turn erasure coding off, regardless of
+    filesystem defaults.  On HDFS, erasure coded files will not update as quickly as regular
+    replicated files, so they application updates will take longer to appear in the History Server.
+    Note that even if this is true, spark will still not force the file to erasure coding, it will
```
[the remark this answers is not in the document]
Comment (kiszk) [reply docs/configuration.md]: nit: `to erasure coding` -> `to use erasure coding`?

**Comment by squito** [comment]
jenkins, retest this please

known flaky: https://issues.apache.org/jira/browse/SPARK-24285

**Review by attilapiros — APPROVED** [review]
LGTM

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala [hunk 2]
```diff
@@ -471,4 +473,42 @@ object SparkHadoopUtil {
...
+   * hflush(), hsync(), or append()
+   * https://hadoop.apache.org/docs/r3.0.0/hadoop-project-dist/hadoop-hdfs/HDFSErasureCoding.html#Limitations
+   */
+  // scalastyle:on line.size.limit
+  def createNonECFile(fs: FileSystem, path: Path): FSDataOutputStream = {
+    try {
+      // Use reflection as this uses apis only avialable in hadoop 3
+      val builderMethod = fs.getClass().getMethod("createFile", classOf[Path])
+      val builder = builderMethod.invoke(fs, path)
+      builderReflection match {
```
[the remark this answers is not in the document]
Comment (xiao-chen) [reply core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala]: FYI - I created https://issues.apache.org/jira/browse/HDFS-14038 for hadoop discussion.

**Comment by vanzin** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala [hunk 3]
```diff
@@ -471,4 +472,33 @@ object SparkHadoopUtil {
...
+      val builderMethod = fs.getClass().getMethod("createFile", classOf[Path])
+      val builder = builderMethod.invoke(fs, path)
+      val builderCls = builder.getClass()
+      // this may throw a NoSuchMethodException if the path is not on hdfs
+      val replicateMethod = builderCls.getMethod("replicate")
+      val buildMethod = builderCls.getMethod("build")
+      val b2 = replicateMethod.invoke(builder)
+      buildMethod.invoke(b2).asInstanceOf[FSDataOutputStream]
+    } catch {
+      case  _: NoSuchMethodException =>
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala]: Any chance of `InvocationTargetException`?

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala]: If you get that exception, it's a proper error that shouldn't fall back to this code path.

File: core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala [hunk 4]
```diff
@@ -471,4 +473,42 @@ object SparkHadoopUtil {
...
+  lazy val builderReflection: Option[(Class[_], Method, Method)] = Try {
+    val cls = Utils.classForName(
+      "org.apache.hadoop.hdfs.DistributedFileSystem$HdfsDataOutputStreamBuilder")
+    (cls, cls.getMethod("replicate"), cls.getMethod("build"))
+  }.toOption
+
+  // scalastyle:off line.size.limit
+  /**
+   * Create a path that uses replication instead of erasure coding, regardless of the default
+   * configuration in hdfs for the given path.  This can be helpful as hdfs ec doesn't support
```
[the remark this answers is not in the document]
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala]: "ec" is already explained in the line above. no need to repeat it.

**Comment by vanzin** [comment]
Merging to master.

File: docs/configuration.md [hunk 5]
```diff
@@ -761,6 +761,17 @@ Apart from these, the following properties are also available, and may be useful
     Compression will use <code>spark.io.compression.codec</code>.
   </td>
 </tr>
+<tr>
+  <td><code>spark.eventLog.allowErasureCoding</code></td>
```
Comment (gatorsmile) [inline docs/configuration.md]: Change `spark.eventLog.allowErasureCoding` to `spark.eventLog.allowErasureCoding.enabled` ?

Comment (gatorsmile) [reply docs/configuration.md]: cc @Ngone51 Could you submit a PR?

Comment (Ngone51) [reply docs/configuration.md]: Sure.

Comment (Ngone51) [reply docs/configuration.md]: Added #26998

**Pull request closed without merging**